README.md
# udp-json <!-- omit in toc -->
<p align="center">
  <img src="https://github.com/Rapsssito/udp-json/workflows/release/badge.svg" />
  <img src="https://img.shields.io/npm/dw/udp-json" />
  <img src="https://img.shields.io/npm/v/udp-json?color=gr&label=npm%20version" />
<p/>

Lightweight JSON UDP socket API written in JavaScript for NodeJS. It provides a simple protocol on top of UDP to send/receive JSON objects. **It does not handle UDP unreliability**.

## Table of Contents <!-- omit in toc -->

- [Protocol](#protocol)
- [Usage](#usage)
- [API](#api)
  - [JSONSocket](#jsonsocket)
    - [`JSONSocket()`](#jsonsocket-1)
    - [`send()`](#send)
    - [`'message-complete'`](#message-complete)
    - [`'message-error'`](#message-error)
    - [`'message-timeout'`](#message-timeout)
- [Maintainers](#maintainers)
- [License](#license)

## Protocol
A UDP datagram consists of a datagram header and a data section. In the data section of each UDP datagram, `udp-json` adds another header and a new data section. The header consists of 3 fields, each of which is 4 bytes (32 bits): 

```
+----------+----------------------+------------------------+
| ID (4 B) | Datagram count (4 B) | Current datagram (4 B) |
+----------+----------------------+------------------------+
|                                                          |
|                     Partial Data                         |
|                                                          |
+----------+----------------------+------------------------+
```

### ID <!-- omit in toc -->
This field identifies to which data refers this partial data.

### Datagram count <!-- omit in toc -->
This field specifies the number of UDP datagrams needed to complete the data.

### Current datagram <!-- omit in toc -->
This field identifies this UDP datagram location inside all the UDP datagrams to complete the data in the correct order.

### Partial Data <!-- omit in toc -->
Fragment of JSON stringified data encoded in `utf-8`.

### Example <!-- omit in toc -->
Example of sending the object `{ attribute: "dummy" }` split in two datagrams.
```
+----------+----------------------+------------------------+
| ID = 333 | Datagram count = 2   | Current datagram = 0   |
+----------+----------------------+------------------------+
|                                                          |
|                         {attr                            |
|                                                          |
+----------+----------------------+------------------------+

+----------+----------------------+------------------------+
| ID = 333 | Datagram count = 2   | Current datagram = 1   |
+----------+----------------------+------------------------+
|                                                          |
|                    ibute: "dummy"}                       |
|                                                          |
+----------+----------------------+------------------------+
```

## Usage
See [examples](examples/) folder for more info.
```javascript
const JSONSocket = require('udp-json');

// Listener socket
const socket = dgram.createSocket('udp4');
socket.bind(56554, '127.0.0.1');
const jsonSocket = new JSONSocket(socket)
jsonSocket.on('message-complete', (msg, rinfo) => {
    console.log('Message received', rinfo, msg);
})

// Sender socket
const socket2 = dgram.createSocket('udp4');
const jsonSocket2 = new JSONSocket(socket2, {maxPayload: 496, timeout: 1000});
jsonSocket2.send({ dummy: 'Dummy Text' }, 56554, '127.0.0.1', (e) => {
    if (e) {
        console.log('error', e);
        return;
    }
    console.log('Message sent');
});
```

## API
Here are listed all methods implemented in `udp-json`.

### JSONSocket
* **Methods:**
  * [`JSONSocket(Socket[, options])`](#jsonsocket)
  * [`send(obj, port[, address][, callback])`](#send)
* **Events:**
  * [`'message-complete'`](#message-complete)
  * [`'message-error'`](#message-error)
  * [`'message-timeout'`](#message-timeout)

#### `JSONSocket()`
`JSONSocket(Socket[, options])`. Creates a JSON socket using the given [`options`](#jsonsocket-options).
##### `JSONSocket: options` <!-- omit in toc -->
Available options for creating a socket. It must be an `object` with the following properties:

| Property     | Type       | Description                                                                                                                                                |
| ------------ | ---------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `maxPayload` | `<number>` | Max payload size (in bytes). If the resulting byte array exceeds this size, it will be splitted in chunks sent in different datagrams. **Default**: `496`. |
| `timeout`    | `<number>` | Max time (in ms) between receiving a new datagram and the last datagram received. **Default**: `1000`.                                                     |

#### `send()`
`send(obj, port[, address][, callback])`. Broadcasts an object on the socket. The destination `port` must be specified. An optional `callback` function may be specified to as a way of reporting errors or for determining when all datagrams have been sent.

#### `'message-complete'`
The `'message-complete'` event is emitted when a new object is available on a socket. The event handler function is passed two arguments: `obj` and `rinfo`.

#### `'message-error'`
The `'message-error'` event is emitted whenever any error occurs. The event handler function is passed a single `Error` object.

#### `'message-timeout'`
The `'message-timeout'` event is emitted if a partial message times out from not receiving a new datagram. The event handler function is passed a single `Error` object.

## Maintainers

* [Rapsssito](https://github.com/rapsssito) [[Support me :heart:](https://github.com/sponsors/Rapsssito)]

## License

The library is released under the MIT license. For more information see [`LICENSE`](/LICENSE).


CHANGELOG.md
## [1.2.1](https://github.com/Rapsssito/udp-json/compare/v1.2.0...v1.2.1) (2021-02-01)


### Bug Fixes

* Add NodeJS backwards compatibility ([57c469a](https://github.com/Rapsssito/udp-json/commit/57c469a1bc9bb0a9b2b717a535c93ebb946236c2))

# [1.2.0](https://github.com/Rapsssito/udp-json/compare/v1.1.3...v1.2.0) (2021-02-01)


### Features

* Add EventTypes typing ([0e9f3e7](https://github.com/Rapsssito/udp-json/commit/0e9f3e79eb2d8c062b30da94c3abbf1b5937c9d1))

## [1.1.3](https://github.com/Rapsssito/udp-json/compare/v1.1.2...v1.1.3) (2021-01-27)


### Bug Fixes

* Switch from last datagram to datagram count ([c028809](https://github.com/Rapsssito/udp-json/commit/c02880931552f65ad3c02733428abdab279fcc8a))

## [1.1.2](https://github.com/Rapsssito/udp-json/compare/v1.1.1...v1.1.2) (2021-01-27)


### Bug Fixes

* Switch from total datagrams to last datagram ([7702cd9](https://github.com/Rapsssito/udp-json/commit/7702cd9322cd90f9d0e729e47094ba04a06b5873))

## [1.1.1](https://github.com/Rapsssito/udp-json/compare/v1.1.0...v1.1.1) (2021-01-27)


### Bug Fixes

* Add ready-to-go examples ([127102e](https://github.com/Rapsssito/udp-json/commit/127102eaec53f49f43b0e51dd4de6794138ae9c1))

# [1.1.0](https://github.com/Rapsssito/udp-json/compare/v1.0.0...v1.1.0) (2021-01-27)


### Features

* Add message-timeout event ([e7088ce](https://github.com/Rapsssito/udp-json/commit/e7088cefb372bb24cb7fec4e9d5749e299c3c11c))

# 1.0.0 (2021-01-26)


### Bug Fixes

* Fix module exports ([c9c1c97](https://github.com/Rapsssito/udp-json/commit/c9c1c975a2477f235cab958839f87630754d1279))


### Features

* Basic implementation ([26e0d7f](https://github.com/Rapsssito/udp-json/commit/26e0d7f00fcde062e8459eb0eaf0da7e40d5e66f))


__tests__/index.test.js
const JSONSocket = require('../src/index.js');
const dgram = require('dgram');

jest.mock('dgram');
const mockedDgram = /** @type {jest.Mocked<typeof dgram>} */ (dgram);
// Mock date
Date.now = () => 0;

const PORT = 50050;
const ADDRESS = '0.0.0.0';

/**
 * @param {any} obj
 * @param {number} size
 */
function buildBufferFromObj(obj, size) {
    const objString = JSON.stringify(obj);
    const buff = Buffer.from(objString);
    const chunks = [];
    for (let startPoint = 0; startPoint < buff.length; startPoint += size) {
        const endPoint = startPoint + size > buff.length ? buff.length : startPoint + size;
        chunks.push(buff.slice(startPoint, endPoint));
    }
    const datagramCount = chunks.length;
    const id = (Date.now() & 0xffffffff) >>> 0;
    const resultBuffers = chunks.map((ch, index) => {
        const header = Buffer.allocUnsafe(12);
        // ID
        header.writeUInt32BE(id);
        // Datagram count
        header.writeUInt32BE(datagramCount, 4);
        // Current datagram
        header.writeUInt32BE(index, 8);
        return Buffer.concat([header, ch]);
    });
    return resultBuffers;
}

test('send-default', () => {
    return new Promise((done) => {
        const mockSend = jest.fn((_msg, _port, _address, callback) => callback());
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({ on: jest.fn(), send: mockSend }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const sentObject = { dummy: 'Dummy' };
        jsonSocket.send(sentObject, PORT, ADDRESS, (e) => {
            expect(e).toBeFalsy();
            expect(mockSend).toHaveBeenCalledTimes(1);
            expect(mockSend).toHaveBeenCalledWith(
                buildBufferFromObj(sentObject, 1000)[0],
                PORT,
                ADDRESS,
                expect.anything()
            );
            done(null);
        });
    });
});

test('send-without-callback', () => {
    return new Promise((done) => {
        const mockSend = jest.fn((msg, port, address, callback) => {
            callback();
            expect(msg).not.toBeFalsy();
            expect(port).toBe(PORT);
            expect(address).toBe(ADDRESS);
            done(null);
        });
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({ on: jest.fn(), send: mockSend }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const sentObject = { dummy: 'Dummy' };
        jsonSocket.send(sentObject, PORT, ADDRESS);
    });
});

test('send-error', () => {
    return new Promise((done) => {
        const mockSend = jest.fn((_msg, _port, _address, callback) => callback(new Error('Dummy error')));
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({ on: jest.fn(), send: mockSend }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const sentObject = { dummy: 'Dummy' };
        jsonSocket.send(sentObject, PORT, ADDRESS, (e) => {
            expect(e).not.toBeFalsy();
            expect(mockSend).toHaveBeenCalledTimes(1);
            expect(mockSend).toHaveBeenCalledWith(
                buildBufferFromObj(sentObject, 1000)[0],
                PORT,
                ADDRESS,
                expect.anything()
            );
            done(null);
        });
    });
});

test('send-split-twice', () => {
    return new Promise((done) => {
        const mockSend = jest.fn((_msg, _port, _address, callback) => callback());
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({ on: jest.fn(), send: mockSend }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket, { maxPayload: 12 });
        const sentObject = { dummy: 'Dummy' };
        jsonSocket.send(sentObject, PORT, ADDRESS, (e) => {
            expect(e).toBeFalsy();
            expect(mockSend).toHaveBeenCalledTimes(2);
            done(null);
        });
    });
});

test('message-default', () => {
    return new Promise((done) => {
        /** @type {any} */
        let mockCallback;
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({
            on: (/** @type {string} */ _evt, /** @type {(arg0: Buffer, arg1: any) => void}*/ callback) => {
                mockCallback = callback;
            },
        }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const messageErrorListener = jest.fn();
        jsonSocket.on('message-error', messageErrorListener);
        jsonSocket.on('message-complete', (msg, rinfo) => {
            expect(rinfo).toEqual(rinfoTest);
            expect(msg).toEqual(testJSON);
            expect(messageErrorListener).not.toHaveBeenCalled();
            done(null);
        });
        const rinfoTest = { address: '127.0.0.1', port: 5666 };
        const testJSON = { dummy: 'Dummy' };
        const messageTest = buildBufferFromObj(testJSON, 1000)[0];
        mockCallback(messageTest, rinfoTest);
    });
});

test('message-error', () => {
    return new Promise((done) => {
        /** @type {any} */
        let mockCallback;
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({
            on: (/** @type {string} */ _evt, /** @type {(arg0: Buffer, arg1: any) => void}*/ callback) => {
                mockCallback = callback;
            },
        }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const messageCompleteListener = jest.fn();
        jsonSocket.on('message-complete', messageCompleteListener);
        jsonSocket.on('message-error', (e) => {
            expect(e).not.toBeFalsy();
            expect(messageCompleteListener).not.toHaveBeenCalled();
            done(null);
        });
        const rinfoTest = { address: '127.0.0.1', port: 5666 };
        const testJSON = { dummy: 'Dummy' };
        const messageTest = buildBufferFromObj(testJSON, 1000)[0];
        messageTest[messageTest.length - 1] = 0; // Modify data in transit
        mockCallback(messageTest, rinfoTest);
    });
});

test('message-split-twice', () => {
    return new Promise((done) => {
        /** @type {any} */
        let mockCallback;
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({
            on: (/** @type {string} */ _evt, /** @type {(arg0: Buffer, arg1: any) => void}*/ callback) => {
                mockCallback = callback;
            },
        }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const messageErrorListener = jest.fn();
        jsonSocket.on('message-error', messageErrorListener);
        jsonSocket.on('message-complete', (msg, rinfo) => {
            expect(rinfo).toEqual(rinfoTest);
            expect(msg).toEqual(testJSON);
            expect(messageErrorListener).not.toHaveBeenCalled();
            done(null);
        });
        const rinfoTest = { address: '127.0.0.1', port: 5666 };
        const testJSON = { dummy: 'Dummy' };
        const messagesTest = buildBufferFromObj(testJSON, 12);
        for (const ms of messagesTest) mockCallback(ms, rinfoTest);
    });
});

test('message-timeout', () => {
    return new Promise((done) => {
        /** @type {any} */
        let mockCallback;
        const mockedSocket = jest.fn();
        mockedSocket.mockImplementation(() => ({
            on: (/** @type {string} */ _evt, /** @type {(arg0: Buffer, arg1: any) => void}*/ callback) => {
                mockCallback = callback;
            },
        }));
        mockedDgram.createSocket.mockImplementation(mockedSocket);
        const socket = mockedDgram.createSocket('udp4');
        const jsonSocket = new JSONSocket(socket);
        const messageErrorListener = jest.fn();
        const messageCompleteListener = jest.fn();
        jsonSocket.on('message-error', messageErrorListener);
        jsonSocket.on('message-complete', messageCompleteListener);
        jsonSocket.on('message-timeout', (e) => {
            expect(e).not.toBeFalsy();
            expect(messageErrorListener).not.toHaveBeenCalled();
            expect(messageCompleteListener).not.toHaveBeenCalled();
            done(null);
        });
        const rinfoTest = { address: '127.0.0.1', port: 5666 };
        const testJSON = { dummy: 'Dummy' };
        const messagesTest = buildBufferFromObj(testJSON, 12);
        mockCallback(messagesTest[1], rinfoTest);
        jest.runOnlyPendingTimers();
    });
});


lib/types/index.d.ts
/// <reference types="node" />
export = JSONSocket;
/**
 * @typedef {'message-complete' | 'message-timeout' | 'message-error'} EventTypes
 * @extends {EventEmitter<EventTypes, any>}
 */
declare class JSONSocket extends EventEmitter<EventTypes, any> {
    /**
     * @param {import('dgram').Socket} udpSocket
     * @param {{maxPayload?: number, timeout?: number}} [options]
     */
    constructor(udpSocket: import('dgram').Socket, options?: {
        maxPayload?: number | undefined;
        timeout?: number | undefined;
    } | undefined);
    udpSocket: import("dgram").Socket;
    /** @private @type {Map<string, PendingDatagram>} */
    private idMap;
    maxPayload: number;
    timeout: number;
    /**
     * @param {any} obj
     * @param {number} port
     * @param {string} [address]
     * @param {(error: Error | null) => void} [callback]
     */
    send(obj: any, port: number, address?: string | undefined, callback?: ((error: Error | null) => void) | undefined): void;
    /**
     * @private
     * @param {Buffer} buff
     * @param {number} size byte
     */
    private _getChunks;
    /**
     * @private
     * @param {number} id
     * @param {number} currentDatagram
     * @param {number} datagramCount
     */
    private _buildHeader;
    /**
     * @private
     */
    private _loadListeners;
    /**
     * @private
     * @param {Buffer} msg
     * @param {import('dgram').RemoteInfo} rinfo
     */
    private _onMessage;
    /**
     * @private
     * @param {string} id
     * @param {string} partialPayload
     * @param {import('dgram').RemoteInfo} rinfo
     * @param {Error} error
     */
    private _onDatagramError;
    /**
     * @private
     * @param {string} id
     * @param {string} partialPayload
     * @param {import('dgram').RemoteInfo} rinfo
     * @param {Error} error
     */
    private _onDatagramTimeout;
    /**
     * @private
     * @param {string} id
     * @param {any} payload
     * @param {import('dgram').RemoteInfo} rinfo
     */
    private _onDatagramComplete;
    /**
     * @private
     * @param {Buffer} msg
     * @param {import('dgram').RemoteInfo} rinfo
     */
    private _unwrapMessage;
}
declare namespace JSONSocket {
    export { EventTypes };
}
type EventTypes = "message-complete" | "message-timeout" | "message-error";
import EventEmitter = require("eventemitter3");


src/index.js
const EventEmitter = require('eventemitter3');

class PendingDatagram {
    /**
     * @param {{datagramCount: number, timeout: number}} options
     * @param {import('dgram').RemoteInfo} rinfo
     * @param {(partialPayload: string, error: Error) => void} errorCallback
     * @param {(partialPayload: string, error: Error) => void} timeoutCallback
     * @param {(payload: any) => void} completeCallback
     */
    constructor(options, rinfo, errorCallback, timeoutCallback, completeCallback) {
        this.timeout = options.timeout;
        this.datagramCount = options.datagramCount;
        /** @private @type {Map<number, string>} */
        this.payloadMap = new Map();
        this.rinfo = rinfo;
        this.errorCallback = errorCallback;
        this.timeoutCallback = timeoutCallback;
        this.completeCallback = completeCallback;
        this.currentTimeout = undefined;
    }

    /**
     * @param {number} datagramIndex
     * @param {string} payload
     */
    addPayload(datagramIndex, payload) {
        this.resetTimeout();
        this.payloadMap.set(datagramIndex, payload);
        if (this.isComplete()) {
            this.stopTimeout();
            const rawPayload = this.getPayload();
            try {
                const jsonPayload = JSON.parse(rawPayload);
                this.completeCallback(jsonPayload);
            } catch (e) {
                this.errorCallback(rawPayload, e);
            }
        }
    }

    stopTimeout() {
        if (this.currentTimeout !== undefined) clearTimeout(this.currentTimeout);
    }

    startTimeout() {
        this.currentTimeout = setTimeout(
            () => this.timeoutCallback(this.getPayload(), new Error(`Timeout: ${this.timeout}`)),
            this.timeout
        );
    }

    resetTimeout() {
        this.stopTimeout();
        this.startTimeout();
    }

    isComplete() {
        if (this.payloadMap.size !== this.datagramCount) return false;
        const valueSum = Array.from(this.payloadMap.keys()).reduce((a, b) => a + b, 0);
        const correctSum = ((this.datagramCount - 1) * this.datagramCount) / 2;
        return valueSum === correctSum;
    }

    getPayload() {
        let result = '';
        for (let i = 0; i < this.datagramCount; i++) {
            result += this.payloadMap.has(i) ? this.payloadMap.get(i) : '';
        }
        return result;
    }
}

/**
 * @typedef {'message-complete' | 'message-timeout' | 'message-error'} EventTypes
 * @extends {EventEmitter<EventTypes, any>}
 */
class JSONSocket extends EventEmitter {
    /**
     * @param {import('dgram').Socket} udpSocket
     * @param {{maxPayload?: number, timeout?: number}} [options]
     */
    constructor(udpSocket, options = {}) {
        super();
        this.udpSocket = udpSocket;
        /** @private @type {Map<string, PendingDatagram>} */
        this.idMap = new Map();
        this._loadListeners();
        this.maxPayload = options.maxPayload || 496; // 508 max safe UDP payload - 12 headers bytes = 496
        this.timeout = options.timeout || 1000;
    }

    /**
     * @param {any} obj
     * @param {number} port
     * @param {string} [address]
     * @param {(error: Error | null) => void} [callback]
     */
    send(obj, port, address, callback = () => {}) {
        const objString = JSON.stringify(obj);
        const chunks = this._getChunks(Buffer.from(objString), this.maxPayload);
        const datagramCount = chunks.length;
        const id = (Date.now() & 0xffffffff) >>> 0;
        const sendPromises = chunks.map((ch, index) => {
            return new Promise((resolve, reject) => {
                const msg = Buffer.concat([this._buildHeader(id, index, datagramCount), ch]);
                this.udpSocket.send(msg, port, address, (err) => {
                    if (err) reject(err);
                    else resolve(null);
                });
            });
        });
        Promise.all(sendPromises)
            .then(() => callback(null))
            .catch(callback);
    }

    /**
     * @private
     * @param {Buffer} buff
     * @param {number} size byte
     */
    _getChunks(buff, size) {
        const chunks = [];
        for (let startPoint = 0; startPoint < buff.length; startPoint += size) {
            const endPoint = startPoint + size > buff.length ? buff.length : startPoint + size;
            chunks.push(buff.slice(startPoint, endPoint));
        }
        return chunks;
    }

    /**
     * @private
     * @param {number} id
     * @param {number} currentDatagram
     * @param {number} datagramCount
     */
    _buildHeader(id, currentDatagram, datagramCount) {
        const header = Buffer.allocUnsafe(12);
        // ID
        header.writeUInt32BE(id);
        // Datagram count
        header.writeUInt32BE(datagramCount, 4);
        // Current datagram
        header.writeUInt32BE(currentDatagram, 8);
        return header;
    }

    /**
     * @private
     */
    _loadListeners() {
        this.udpSocket.on('message', (msg, rinfo) => this._onMessage(msg, rinfo));
    }

    /**
     * @private
     * @param {Buffer} msg
     * @param {import('dgram').RemoteInfo} rinfo
     */
    _onMessage(msg, rinfo) {
        const { id, currentDatagram, datagramCount, payload } = this._unwrapMessage(msg, rinfo);
        let pendingDatagram = this.idMap.get(id);
        if (pendingDatagram === undefined) {
            pendingDatagram = new PendingDatagram(
                { datagramCount: datagramCount, timeout: this.timeout },
                rinfo,
                (partialPayload, error) => this._onDatagramError(id, partialPayload, rinfo, error),
                (partialPayload, error) => this._onDatagramTimeout(id, partialPayload, rinfo, error),
                (payload) => this._onDatagramComplete(id, payload, rinfo)
            );
            this.idMap.set(id, pendingDatagram);
        }
        pendingDatagram.addPayload(currentDatagram, payload);
    }

    /**
     * @private
     * @param {string} id
     * @param {string} partialPayload
     * @param {import('dgram').RemoteInfo} rinfo
     * @param {Error} error
     */
    _onDatagramError(id, partialPayload, rinfo, error) {
        const buildError = new Error(`Error ${JSON.stringify(rinfo)} current payload ${partialPayload}: ${error}`);
        this.emit('message-error', buildError);
        this.idMap.delete(id);
    }

    /**
     * @private
     * @param {string} id
     * @param {string} partialPayload
     * @param {import('dgram').RemoteInfo} rinfo
     * @param {Error} error
     */
    _onDatagramTimeout(id, partialPayload, rinfo, error) {
        const buildError = new Error(`Timeout ${JSON.stringify(rinfo)} current payload ${partialPayload}: ${error}`);
        this.emit('message-timeout', buildError);
        this.idMap.delete(id);
    }

    /**
     * @private
     * @param {string} id
     * @param {any} payload
     * @param {import('dgram').RemoteInfo} rinfo
     */
    _onDatagramComplete(id, payload, rinfo) {
        this.emit('message-complete', payload, rinfo);
        this.idMap.delete(id);
    }

    /**
     * @private
     * @param {Buffer} msg
     * @param {import('dgram').RemoteInfo} rinfo
     */
    _unwrapMessage(msg, rinfo) {
        const id = `${rinfo.address}:${rinfo.port}:${msg.toString('base64', 0, 4)}`;
        const datagramCount = msg.readInt32BE(4);
        const currentDatagram = msg.readInt32BE(8);
        const payload = msg.toString('utf8', 12);
        return { id, currentDatagram, datagramCount, payload };
    }
}

module.exports = JSONSocket;
